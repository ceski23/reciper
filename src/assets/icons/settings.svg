<svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none">
    <path
        d="M19.4322 12.98C19.4722 12.66 19.5022 12.34 19.5022 12C19.5022 11.66 19.4722 11.34 19.4322 11.02L21.5422 9.37C21.7322 9.22 21.7822 8.95 21.6622 8.73L19.6622 5.27C19.5722 5.11 19.4022 5.02 19.2222 5.02C19.1622 5.02 19.1022 5.03 19.0522 5.05L16.5622 6.05C16.0422 5.65 15.4822 5.32 14.8722 5.07L14.4922 2.42C14.4622 2.18 14.2522 2 14.0022 2H10.0022C9.75221 2 9.54221 2.18 9.51221 2.42L9.13221 5.07C8.52221 5.32 7.96221 5.66 7.44221 6.05L4.95221 5.05C4.89221 5.03 4.83221 5.02 4.77221 5.02C4.60221 5.02 4.43221 5.11 4.34221 5.27L2.34221 8.73C2.21221 8.95 2.27221 9.22 2.46221 9.37L4.57221 11.02C4.53221 11.34 4.50221 11.67 4.50221 12C4.50221 12.33 4.53221 12.66 4.57221 12.98L2.46221 14.63C2.27221 14.78 2.22221 15.05 2.34221 15.27L4.34221 18.73C4.43221 18.89 4.60221 18.98 4.78221 18.98C4.84221 18.98 4.90221 18.97 4.95221 18.95L7.44221 17.95C7.96221 18.35 8.52221 18.68 9.13221 18.93L9.51221 21.58C9.54221 21.82 9.75221 22 10.0022 22H14.0022C14.2522 22 14.4622 21.82 14.4922 21.58L14.8722 18.93C15.4822 18.68 16.0422 18.34 16.5622 17.95L19.0522 18.95C19.1122 18.97 19.1722 18.98 19.2322 18.98C19.4022 18.98 19.5722 18.89 19.6622 18.73L21.6622 15.27C21.7822 15.05 21.7322 14.78 21.5422 14.63L19.4322 12.98ZM17.4522 11.27C17.4922 11.58 17.5022 11.79 17.5022 12C17.5022 12.21 17.4822 12.43 17.4522 12.73L17.3122 13.86L18.2022 14.56L19.2822 15.4L18.5822 16.61L17.3122 16.1L16.2722 15.68L15.3722 16.36C14.9422 16.68 14.5322 16.92 14.1222 17.09L13.0622 17.52L12.9022 18.65L12.7022 20H11.3022L10.9522 17.52L9.89221 17.09C9.46221 16.91 9.06221 16.68 8.66221 16.38L7.75221 15.68L6.69221 16.11L5.42221 16.62L4.72221 15.41L5.80221 14.57L6.69221 13.87L6.55221 12.74C6.52221 12.43 6.50221 12.2 6.50221 12C6.50221 11.8 6.52221 11.57 6.55221 11.27L6.69221 10.14L5.80221 9.44L4.72221 8.6L5.42221 7.39L6.69221 7.9L7.73221 8.32L8.63221 7.64C9.06221 7.32 9.47221 7.08 9.88221 6.91L10.9422 6.48L11.1022 5.35L11.3022 4H12.6922L13.0422 6.48L14.1022 6.91C14.5322 7.09 14.9322 7.32 15.3322 7.62L16.2422 8.32L17.3022 7.89L18.5722 7.38L19.2722 8.59L18.2022 9.44L17.3122 10.14L17.4522 11.27ZM12.0022 8C9.79221 8 8.00221 9.79 8.00221 12C8.00221 14.21 9.79221 16 12.0022 16C14.2122 16 16.0022 14.21 16.0022 12C16.0022 9.79 14.2122 8 12.0022 8ZM12.0022 14C10.9022 14 10.0022 13.1 10.0022 12C10.0022 10.9 10.9022 10 12.0022 10C13.1022 10 14.0022 10.9 14.0022 12C14.0022 13.1 13.1022 14 12.0022 14Z"
        fill="currentColor" />
</svg>