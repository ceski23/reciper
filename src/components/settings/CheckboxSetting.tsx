/* eslint-disable react/require-default-props */
import styled from '@emotion/styled';
import React from 'react';

const InfoContainer = styled.div`
  display: flex;
  flex-direction: column;
  flex: 1;
`;

const Container = styled.label`
  display: flex;
  flex-direction: row;
  margin-bottom: 30px;
`;

const CheckboxInput = styled.input`
  appearance: none;
  width: 20px;
  height: 20px;
  border-radius: 5px;
  border: 4px solid ${(props) => props.theme.colors.background};
  background-color: ${(props) => props.theme.colors.background};
  align-self: center;
  box-shadow: 0 0 1px ${(props) => props.theme.colors.shadow};

  &:checked {
    background-color: ${(props) => props.theme.colors.textalt};
  }
`;

const Hint = styled.p`
  font-size: 12px;
  line-height: 1.4;
  margin-top: -10px;
  color: ${(props) => props.theme.colors.textalt};
  margin-bottom: 0;
`;

const Label = styled.h3`
  margin-top: 0;
`;

interface Props {
  title?: string;
  hint?: string;
  name: string;
  onChange: (value: boolean) => void;
  checked: boolean;
}

export const CheckboxSetting = ({
  name, onChange, checked, title, hint,
}: Props) => (
  <Container>
    <InfoContainer>
      {title && <Label>{title}</Label>}
      {hint && <Hint>{hint}</Hint>}
    </InfoContainer>
    <CheckboxInput type="checkbox" name={name} checked={checked} onChange={(event) => onChange(event.currentTarget.checked)} />
  </Container>
);
