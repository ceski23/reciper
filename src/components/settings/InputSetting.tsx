/* eslint-disable react/require-default-props */
import styled from '@emotion/styled/macro';

import { color } from 'utils/styles/theme';

const InfoContainer = styled.div`
  display: flex;
  flex-direction: column;
  flex: 1;
`;

const Container = styled.label`
  display: flex;
  flex-direction: row;
  margin-bottom: 30px;
`;

const Input = styled.input`
  appearance: none;
  width: 70px;
  height: 40px;
  border-radius: 5px;
  border: 4px solid ${color('background')};
  background-color: ${color('background')};
  align-self: center;
  box-shadow: 0 0 1px ${color('shadow')};
`;

const Hint = styled.p`
  font-size: 12px;
  line-height: 1.4;
  margin-top: -10px;
  color: ${color('textalt')};
  margin-bottom: 0;
`;

const Label = styled.h3`
  margin-top: 0;
`;

interface Props {
  title?: string;
  hint?: string;
  onChange: (value: string) => void;
  value: string;
  type: string;
}

export const InputSetting = ({
  onChange, value, title, hint, type,
}: Props) => (
  <Container>
    <InfoContainer>
      {title && <Label>{title}</Label>}
      {hint && <Hint>{hint}</Hint>}
    </InfoContainer>
    <Input
      type={type}
      min={0}
      value={value}
      onChange={(event) => onChange(event.currentTarget.value)}
    />
  </Container>
);
