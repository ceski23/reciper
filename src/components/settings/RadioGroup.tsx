/* eslint-disable react/require-default-props */
import styled from '@emotion/styled';

import { color } from 'utils/styles/theme';

const RadioContainer = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  gap: 20px;
  background-color: ${color('backgroundHover')};
  border-radius: 20px;
  padding: 20px;
`;

const Container = styled.div`
  display: flex;
  flex-direction: column;
  margin-bottom: 30px;
`;

const RadioInput = styled.input`
  appearance: none;
  width: 20px;
  height: 20px;
  border-radius: 50%;
  border: 4px solid ${color('background')};
  background-color: ${color('background')};
  margin-right: 10px;
  box-shadow: 0 0 1px ${color('shadow')};

  &:checked {
    background-color: ${color('textalt')};
  }
`;

const RadioLabel = styled.label`
  display: flex;
  flex-direction: row;
  align-items: flex-end;
`;

const Hint = styled.p`
  font-size: 12px;
  line-height: 1.4;
  margin-top: -5px;
  margin-bottom: 20px;
  color: ${color('textalt')};
`;

interface Props<T extends string | number> {
  title?: string;
  hint?: string;
  name: string;
  onSelected: (value: T) => void;
  options: Array<{
    text: string,
    value: T
  }>;
  value: T | undefined;
}

// eslint-disable-next-line @typescript-eslint/comma-dangle
export const RadioGroup = <T extends string | number>({
  name, onSelected, options, value, title, hint,
}: Props<T>) => (
  <Container>
    {title && <h3>{title}</h3>}

    {hint && <Hint>{hint}</Hint>}

    <RadioContainer>
      {options.map((option) => (
        // eslint-disable-next-line jsx-a11y/label-has-associated-control
        <RadioLabel key={option.value}>
          <RadioInput
            type="radio"
            name={name}
            value={option.value}
            checked={value === option.value}
            onChange={() => onSelected(option.value)}
          />
          {option.text}
        </RadioLabel>
      ))}
    </RadioContainer>
  </Container>
  );
