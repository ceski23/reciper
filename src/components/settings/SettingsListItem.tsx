import { css, Theme } from '@emotion/react';
import styled from '@emotion/styled';
import React, { FC } from 'react';

import { Styleable } from 'types';

import { color } from 'utils/styles/theme';

const Container = styled.div`
  background-color: ${color('backgroundAlt')};
  border-radius: 10px;
  padding: 26px 20px;
  font-size: 16px;
  font-weight: 600;
  display: flex;
  flex-direction: row;
  align-items: center;
  transition: background-color 0.3s;

  &:hover {
    background-color: ${color('backgroundAltHover')};
  }
`;

const Text = styled.p`
  color: currentColor;
  margin: 0;
`;

export const iconStyles = ({ theme }: { theme: Theme }) => css`
  margin-right: 20px;
  width: 25px;
  fill: ${theme.colors.textalt};
`;

interface Props {
  icon?: React.FunctionComponent<React.SVGProps<SVGSVGElement>> | string;
  text?: string;
  onClick?: () => void;
}

export const SettingsListItem: FC<Props & Styleable> = ({
  icon: Icon, text, children, onClick, className,
}) => (
  <Container onClick={onClick} className={className}>
    {Icon && (typeof Icon === 'string' ? (
      <img src={Icon} alt="" css={(theme) => iconStyles({ theme })} />
    ) : (
      <Icon css={(theme) => iconStyles({ theme })} />
    ))}
    {text ? <Text>{text}</Text> : children}
  </Container>
);
