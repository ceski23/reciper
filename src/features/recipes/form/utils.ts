import { group } from 'radash'
import { type RecipeFormValues } from 'features/recipes/form/scheme'
import { type Recipe } from 'features/recipes/types'

export const mapFormValuesToRecipe = (values: RecipeFormValues) => ({
	...values,
	ingredients: values.ingredients.flatMap(({ items, group }) =>
		items.map(item => ({
			...item,
			group,
		}))
	),
	instructions: values.instructions.flatMap(({ items, group }) =>
		items.map(item => ({
			...item,
			group,
		}))
	),
	tags: values.tags.map(tag => tag.text),
} satisfies Omit<Recipe, 'id' | 'addedDate'>)

export const mapRecipeToFormValues = (recipe: Recipe) => ({
	...recipe,
	ingredients: Object.entries(group(recipe.ingredients, item => item.group ?? '')).map(([group, items]) => ({
		group: group || undefined,
		items: items?.map(({
			// eslint-disable-next-line @typescript-eslint/no-unused-vars
			group,
			...item
		}) => item) ?? [],
	})),
	instructions: Object.entries(group(recipe.instructions, item => item.group ?? '')).map(([group, items]) => ({
		group: group || undefined,
		items: items?.map(({
			// eslint-disable-next-line @typescript-eslint/no-unused-vars
			group,
			...item
		}) => item) ?? [],
	})),
	tags: recipe.tags.map(tag => ({ text: tag })),
} satisfies RecipeFormValues)
