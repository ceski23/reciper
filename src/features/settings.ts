/* eslint-disable max-len */
/* eslint-disable no-param-reassign */
import { Theme } from '@emotion/react/macro';
import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { persistReducer } from 'redux-persist';
import storage from 'redux-persist/lib/storage';
import { RootState } from 'store';

interface SettingsState {
  theme: Theme['type'] | 'system';
  dynamicPrimaryColor: boolean;
  useUnitsConversion: boolean;
  unitsConversion: Record<string, string | undefined>;
  conversionPrecision: number;
}

export const themeName = (theme: SettingsState['theme']) => {
  switch (theme) {
    case 'dark': return 'Ciemny';
    case 'light': return 'Jasny';
    case 'system': return 'Systemowy';
    default:
      throw Error('Unknown theme');
  }
};

const initialState: SettingsState = {
  theme: 'system',
  dynamicPrimaryColor: true,
  useUnitsConversion: true,
  unitsConversion: {},
  conversionPrecision: 2,
};

const slice = createSlice({
  name: 'settings',
  initialState,
  reducers: {
    setTheme: (state, { payload }: PayloadAction<SettingsState['theme']>) => {
      state.theme = payload;
    },
    setDynamicPrimaryColor: (state, { payload }: PayloadAction<SettingsState['dynamicPrimaryColor']>) => {
      state.dynamicPrimaryColor = payload;
    },
    setUseUnitsConversion: (state, { payload }: PayloadAction<SettingsState['useUnitsConversion']>) => {
      state.useUnitsConversion = payload;
    },
    setUnitConversion: (state, { payload }: PayloadAction<{ unit: string, conversion?: string }>) => {
      state.unitsConversion[payload.unit] = payload.conversion;
    },
    setConversionPrecision: (state, { payload }: PayloadAction<SettingsState['conversionPrecision']>) => {
      state.conversionPrecision = payload;
    },
  },
});

export const {
  setTheme, setDynamicPrimaryColor, setUseUnitsConversion, setUnitConversion, setConversionPrecision,
} = slice.actions;

export const selectCurrentThemeType = (state: RootState) => state.settings.theme;

export const selectDynamicPrimaryColor = (state: RootState) => state.settings.dynamicPrimaryColor;

export const selectUnitsConversions = (state: RootState) => state.settings.unitsConversion;

export const selectUseUnitsConversion = (state: RootState) => state.settings.useUnitsConversion;

export const selectConversionPrecision = (state: RootState) => state.settings.conversionPrecision;

export default persistReducer({
  key: 'settings',
  storage,
}, slice.reducer);
