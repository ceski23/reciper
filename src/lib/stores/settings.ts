import { createStore } from 'stan-js'
import { storage } from 'stan-js/storage'

export type Settings = {
	theme: {
		colorScheme: 'dark' | 'light' | undefined
		dynamicColor: boolean
		disabledAnimations: boolean
	}
	language: string | undefined
	account: {
		sync: boolean
	}
}

export const settingsStore = createStore<Settings>({
	theme: storage({
		colorScheme: undefined,
		dynamicColor: true,
		disabledAnimations: false,
	}),
	language: storage(undefined),
	account: storage({
		sync: false,
	}),
})
