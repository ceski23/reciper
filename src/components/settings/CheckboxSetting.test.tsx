import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { renderWithProviders } from 'test/utils';

import { CheckboxSetting } from 'components/settings/CheckboxSetting';

describe('CheckboxSetting', () => {
  it('should render correctly', () => {
    const { asFragment } = renderWithProviders(
      <CheckboxSetting
        title="Test"
        name="test1"
        checked
        onChange={vi.fn()}
        hint="test2"
      />,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should change value', async () => {
    const handleChange = vi.fn();
    const user = userEvent.setup();

    renderWithProviders(
      <CheckboxSetting
        name="test"
        checked
        onChange={handleChange}
        title="Test"
      />,
    );

    const checkbox = screen.getByRole('checkbox', { name: /test/i });
    await user.click(checkbox);
    expect(handleChange).toBeCalledTimes(1);
  });
});
