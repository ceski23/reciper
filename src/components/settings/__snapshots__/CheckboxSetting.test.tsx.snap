// Vitest Snapshot v1

exports[`CheckboxSetting > should render correctly 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  margin-bottom: 30px;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.emotion-4 {
  margin-top: 0;
}

.emotion-6 {
  font-size: 12px;
  line-height: 1.4;
  margin-top: -10px;
  color: #365463;
  margin-bottom: 0;
}

.emotion-8 {
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  width: 20px;
  height: 20px;
  border-radius: 5px;
  border: 4px solid #f6f8f9;
  background-color: #f6f8f9;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  box-shadow: 0 0 1px rgba(57,66,70,0.3);
}

.emotion-8:checked {
  background-color: #365463;
}

<label
    class="emotion-0 emotion-1"
  >
    <div
      class="emotion-2 emotion-3"
    >
      <h3
        class="emotion-4 emotion-5"
      >
        Test
      </h3>
      <p
        class="emotion-6 emotion-7"
      >
        test2
      </p>
    </div>
    <input
      checked=""
      class="emotion-8 emotion-9"
      name="test1"
      type="checkbox"
    />
  </label>
</DocumentFragment>
`;
